spec/data-model.md

```diff
@@ -0,0 +1,177 @@
+
+To ensure compatibility across all platforms,
+this interchange data model is defined in terms of JSON-compatible values.
+While this document uses TypeScript syntax for their definition,
+the canonical and authoritative source is the `message.json`
+[JSON Schema](https://json-schema.org/) definition.
+
+## Messages
+
+A `SelectMessage` corresponds to a syntax message that includes _selectors_.
+A message without _selectors_ and with a single _pattern_ is represented by a `PatternMessage`.
+
+```ts
+type Message = PatternMessage | SelectMessage
+
+interface PatternMessage {
+  type: 'message'
+  declarations: Declaration[]
+  pattern: Pattern
+}
+
+interface SelectMessage {
+  type: 'select'
+  declarations: Declaration[]
+  selectors: Expression[]
+  variants: Variant[]
+}
+```
+
+Each message _declaration_ is represented by a `Declaration`,
+which connects the left-hand side `target` _variable_
+with its right-hand side `value`.
+
+```ts
+interface Declaration {
+  target: VariableRef
+  value: Expression
+}
+```
+
+In a `SelectMessage`, the `keys` and `value` of each _variant_ are kept as an ordered list.
+For the `CatchallKey`, a string `value` may be provided to retain an identifier.
+This is always `'*'` in MessageFormat 2 syntax, but may vary in other formats.
+
+```ts
+interface Variant {
+  keys: Array<Literal | CatchallKey>
+  value: Pattern
+}
+
+interface CatchallKey {
+  type: '*'
+  value?: string
+}
+```
+
+## Patterns
+
+Each `Pattern` represents a single linear _pattern_ without selectors,
+with a `body` made up of `Text` and `Expression` shapes.
+`Text` represents literal _text_,
+while `Expression` wraps each of the potential _expression_ shapes.
+The `value` of `Text` is the "cooked" value (i.e. escape sequences are processed).
+
+Implementations MUST NOT rely on the set of `Expression` `body` values being exhaustive,
+as future versions of this specification MAY define additional expressions.
+If encountering a `body` with an unrecognised value,
+an implementation SHOULD treat it as it would a `Reserved` value.
+
+```ts
+interface Pattern {
+  body: Array<Text | Expression>
+}
+
+interface Text {
+  type: 'text'
+  value: string
+}
+
+interface Expression {
+  type: 'expression'
+  body: Literal | VariableRef | FunctionRef | Reserved
+}
+```
+
+## Expressions
+
+The `Literal` and `VariableRef` correspond to the the _literal_ and _variable_ syntax rules.
+When they are used as the `body` of an `Expression`,
+they represent _expression_ values with no _annotation_.
+
+An _unquoted_ value is represented by a `Literal` with `quoted: false`,
+while a _quoted_ value would have `quoted: true`.
+The `value` of `Literal` is the "cooked" value (i.e. escape sequences are processed).
+
+In a `VariableRef`, the `name` does not include the initial `$` of the _variable_.
+
+```ts
+interface Literal {
+  type: 'literal'
+  quoted: boolean
+  value: string
+}
+
+interface VariableRef {
+  type: 'variable'
+  name: string
+}
+```
```

Comment:
I'm not immediately convinced of the benefits of defining an atomic/simple vs. compound split, as in practice this would be splitting a set of three things into two different categories, at the cost of introducing novel terminology.

What would we gain from this?

Reply:
I thought it would be better-organized that way, but it's true that it's a relatively small set of things. In the interest of not increasing the amount of work left to do, I'll withdraw this suggestion.